sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuFlatMapCoGroupsInPandasExec.scala

```diff
@@ -0,0 +1,137 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.execution.python
+
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.python.PythonWorkerSemaphore
+
+import org.apache.spark.TaskContext
+import org.apache.spark.api.python.{ChainedPythonFunctions, PythonEvalType}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.physical.{AllTuples, ClusteredDistribution,
+  Distribution, Partitioning}
+import org.apache.spark.sql.execution.{BinaryExecNode, CoGroupedIterator, SparkPlan}
+import org.apache.spark.sql.execution.python.{CoGroupedArrowPythonRunner,
+  FlatMapCoGroupsInPandasExec}
+import org.apache.spark.sql.execution.python.rapids.GpuPandasUtils._
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.ArrowUtils
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+
+class GpuFlatMapCoGroupsInPandasExecMeta(
+    flatPandas: FlatMapCoGroupsInPandasExec,
+    conf: RapidsConf,
+    parent: Option[RapidsMeta[_, _, _]],
+    rule: ConfKeysAndIncompat)
+  extends SparkPlanMeta[FlatMapCoGroupsInPandasExec](flatPandas, conf, parent, rule) {
+
+  override def couldReplaceMessage: String = "could partially run on GPU"
+  override def noReplacementPossibleMessage(reasons: String): String =
+    s"cannot run even partially on the GPU because $reasons"
+
+  // Ignore the expressions since columnar way is not supported yet
+  override val childExprs: Seq[BaseExprMeta[_]] = Seq.empty
+
+  override def convertToGpu(): GpuExec =
+    GpuFlatMapCoGroupsInPandasExec(
+      flatPandas.leftGroup, flatPandas.rightGroup,
+      flatPandas.func,
+      flatPandas.output,
+      childPlans.head.convertIfNeeded(), childPlans(1).convertIfNeeded()
+    )
+}
+
+/**
+ *
+ * This GpuFlatMapCoGroupsInPandasExec aims at supporting running Pandas functional code
+ * on GPU at Python side.
+ *
+ * (Currently it will not run on GPU itself, since the columnar way is not implemented yet.)
+ *
+ */
+case class GpuFlatMapCoGroupsInPandasExec(
+    leftGroup: Seq[Attribute],
+    rightGroup: Seq[Attribute],
+    func: Expression,
+    output: Seq[Attribute],
+    left: SparkPlan,
+    right: SparkPlan)
+  extends SparkPlan with BinaryExecNode with GpuExec {
+
+  override def supportsColumnar = false
+  override def doExecuteColumnar(): RDD[ColumnarBatch] = {
+    // TBD
+    super.doExecuteColumnar()
```

Comment:
Same comment as before remove the TBD and possibly just throw an exception yourself.

Reply:
Good idea. Updated